static/tutorial/exception-safety.md

```diff
@@ -0,0 +1,341 @@
+an `IO` action, pure code, during a monadic bind, etc. In our
+`withFile` example above, an async exception can be thrown:
+
+* Before the call to `openFile`
+* During the call to `openFile`
+* Between `openFile` and `finally`
+* After `finally` before exiting `withFile`
+* Inside `inner`
+
+As an exercise, I recommend that you step through these different
+cases and see whether our `withFile` leaks resources or not.
+
+The first trick we need to add to our arsenal is *masking*. This sets
+up a region of code wherein asynchronous exceptions are guaranteed
+_not_ to be thrown. Usage is pretty straightforward:
+
+```haskell
+withFile fp inner = mask $ \restore -> do
+    h <- openFile fp
+    restore (inner h) `finally` closeFile h
+```
+
+The `restore` function will restore the original masking state, which
+usually means "turn async exceptions back on." We restore async
+exceptions for the inner action so as to avoid creating an
+impossible-to-kill thread, but make sure that we do all of our
+allocation and cleanup in a masked state so we can avoid getting
+exceptions at unexpected places. We now know for certain that only
+`IO` actions have the potential to throw an exception, and ensure that
+our code does not leak in the case of any of those actions leaking
+exceptions.
+
+__Exercise__ Write a function called `with2Files` which will open and
+guarantee the close of two files instead of just one. You can first
+implement in terms of `withFile`, and then implement without the help
+of `withFile`.
+
+__Solution__
+
+```haskell
+with2Files fp1 fp2 inner = mask $ \restore -> do
+    h1 <- openFile fp1
+    h2 <- openFile fp2 `onException` closeFile h1
+    restore (inner h1 h2) `finally` closeFile h2 `finally` closeFile h1
+```
+
+The big takeaway you should get from this exercise: while masking
+prevent asynchronous exceptions from running, it does _nothing_ to
+prevent synchronous exceptions. You must still deal with those
+explicitly.
+
+## Exceptions in an allocation function
+
+It's entirely possible that the `openFile` function itself may throw
+an exception (e.g., permission denied to open the given file). If we
+think at all about our defined guarantees above, the behavior of
+`openFile` in that situation should be obvious: it must not leak any
+resources! We must be guaranteed that, if an exception is thrown in
+`openFile`, it cleans up after itself completely.
+
+Using our example of `with2Files`, we can demonstrate the proper way
+to write an `open2Files` function that can be used for allocation:
+
+```haskell
+open2Files fp1 fp2 = do
+    h1 <- openFile fp1
+    h2 <- openFile fp2 `onException` closeFile h1
+    return (h1, h2)
+```
+
+Notice how `h1` will be closed if `fp2` throws an exception, and only
+on success will both handles remain open. This is the behavior desired
+of a well written allocation function.
+
+The other thing to point out is that _we don't mask exceptions_. It is
```

Comment:
I'll push a commit now addressing the first bullet. Regarding the second: the point of this is that `cancel` needn't be mask/uninterruptible in and of itself to be a valid cleanup function, since `cancel` can rely on its caller to use `uninterruptibleMask`.